crates/interpreter/src/exec.rs

```diff
@@ -1055,15 +1067,18 @@ impl<'m> Thread<'m> {
     fn exit_label(mut self) -> ThreadResult<'m> {
         let frame = self.frame();
         let label = frame.labels.pop().unwrap();
+        let popped_values =
+            frame.labels_values.split_off(frame.labels_values.len() - label.values_cnt);
         if frame.labels.is_empty() {
             let frame = self.frames.pop().unwrap();
-            debug_assert_eq!(label.values.len(), frame.arity);
+            debug_assert_eq!(label.values_cnt, frame.arity);
             if self.frames.is_empty() {
-                return ThreadResult::Done(label.values);
+                return ThreadResult::Done(popped_values);
             }
             unsafe { self.parser.restore(frame.ret) };
         }
-        self.values().extend_from_slice(&label.values);
+        self.values().extend(popped_values);
```

Comment:
This can also be optimized out. It only needs to be done when changing frames (until that part is also handled).